Amazon.Ion.ObjectMapper/IonSerializer.cs

```diff
@@ -286,84 +270,88 @@ public object Deserialize(IIonReader reader, Type type)
+                        deserialized = new IonLongSerializer().Deserialize(reader);
+                    }
+                    else
+                    {
+                        deserialized = new IonIntSerializer().Deserialize(reader);
+                    }
+                    break;
+                case IonType.Float:
+                    if (reader.GetTypeAnnotations().Any(s => s.Equals(IonFloatSerializer.ANNOTATION)))
+                    {
+                        deserialized = new IonFloatSerializer().Deserialize(reader);
+                    }
+                    else
+                    {
+                        deserialized = new IonDoubleSerializer().Deserialize(reader);
+                    }
+                    break;
+                case IonType.Decimal:
+                    if (reader.GetTypeAnnotations().Any(s => s.Equals(IonDecimalSerializer.ANNOTATION)))
+                    {
+                        deserialized = new IonDecimalSerializer().Deserialize(reader);
+                    }
+                    else
+                    {
+                        deserialized = new IonBigDecimalSerializer().Deserialize(reader);
+                    }
+                    break;
+                case IonType.String:
+                    deserialized = new IonStringSerializer().Deserialize(reader);
+                    break;
+                case IonType.Symbol:
+                    deserialized = new IonSymbolSerializer().Deserialize(reader);
+                    break;
+                case IonType.Timestamp:
+                    deserialized = new IonDateTimeSerializer().Deserialize(reader);
+                    break;
+                case IonType.Blob:
+                    if (reader.GetTypeAnnotations().Any(s => s.Equals(IonGuidSerializer.ANNOTATION))
+                        || type.IsAssignableTo(typeof(Guid)))
+                    {
+                        deserialized = new IonGuidSerializer(options).Deserialize(reader);
+                    }
+                    else
+                    {
+                        deserialized = new IonByteArraySerializer().Deserialize(reader);
+                    }
+                    break;
+                case IonType.Clob:
+                    deserialized = new IonClobSerializer().Deserialize(reader);
+                    break;
+                case IonType.List:
+                    deserialized = NewIonListSerializer(type).Deserialize(reader);
+                    break;
+                case IonType.Struct:
+                    deserialized = new IonObjectSerializer(this, options, type).Deserialize(reader);
+                    break;
+                default:
+                    throw new NotSupportedException("Don't know how to Deserialize this Ion data. Last IonType was: " + ionType);
             }
 
-            throw new NotSupportedException("Don't know how to Deserialize this Ion data. Last IonType was: " + ionType);
+            currentDepth--;

```

Comment:
If you used the try/finally construct, you also would be able to retain the direct return instead of assigning to `deserialized`.

Reply:
Done.